applications/ShapeOptimizationApplication/python_scripts/algorithm_trust_region.py

```diff
@@ -0,0 +1,682 @@
+        else:
+            obj_id = self.specified_objectives[0]["identifier"].GetString()
+            current_obj_val = self.communicator.getStandardizedValue(obj_id)
+            obj_history = self.data_logger.GetValueHistory(obj_id+"_standardized")
+            step_history = self.data_logger.GetValueHistory("step_length")
+
+            # Check for osciallation
+            objective_is_oscillating = False
+            is_decrease_1 = (current_obj_val - obj_history[self.opt_iteration-1])< 0
+            is_decrease_2 = (obj_history[self.opt_iteration-1] - obj_history[self.opt_iteration-2])<0
+            is_decrease_3 = (current_obj_val - obj_history[self.opt_iteration-3])< 0
+            if (is_decrease_1 and is_decrease_2== False and is_decrease_3) or (is_decrease_1== False and is_decrease_2 and is_decrease_3==False):
+                objective_is_oscillating = True
+
+            # Reduce step length if certain conditions are fullfilled
+            if objective_is_oscillating:
+                return step_history[self.opt_iteration-1]*self.algorithm_settings["step_length_reduction_factor"].GetDouble()
+            else:
+                return step_history[self.opt_iteration-1]
+
+    # --------------------------------------------------------------------------
+    def __ExpressInStepLengthUnit(self, len_obj, len_eqs, len_ineqs, step_length):
+        len_bar_obj = 1/step_length * len_obj
+        len_bar_eqs = ScalarVectorProduct(1/step_length, len_eqs)
+        len_bar_ineqs = ScalarVectorProduct(1/step_length, len_ineqs)
+        return len_bar_obj, len_bar_eqs, len_bar_ineqs
+
+    # --------------------------------------------------------------------------
+    def __DetermineStep(self, len_obj, dir_obj, len_eqs, dir_eqs, len_ineqs, dir_ineqs):
+        print("\n> Starting determination of step...")
+
+        timer = Timer()
+        timer.StartTimer()
+
+        # Create projector object wich can do the projection in the orthogonalized subspace
+        projector = Projector(len_obj, dir_obj, len_eqs, dir_eqs, len_ineqs, dir_ineqs, self.algorithm_settings)
+
+        # 1. Test projection if there is room for objective improvement
+        # I.e., the actual step length to become feasible for an inactive threshold is smaller than 1 and hence a part of the step can be dedicated to objective improvement
+        nargout = 2
+        len_obj_test = 0.01
+        inactive_threshold = 100
+        test_norm_dX, is_projection_sucessfull = projector.RunProjection(len_obj_test, inactive_threshold, nargout)
+
+        print("> Time needed for one projection step = ", timer.GetTotalTime(), "s")
+
+        # 2. Determine step following two different modes depending on the previos found step length to the feasible domain
+        if is_projection_sucessfull:
+            if test_norm_dX < 1: # Minimizing mode
+                print ("\n> Computing projection case 1...")
```

Comment:
printing "Minimizing mode" will make the log more readable

Reply:
I agree. But this is implicitly done by logging the "test_norm_dX". On top of that one directly gets the value as an additional info.